src/stream/src/executor/backfill/snapshot_backfill.rs

```diff
@@ -339,61 +323,41 @@ struct StateOfConsumingSnapshot {
 }
 
 impl UpstreamBufferState for StateOfConsumingSnapshot {
-    async fn is_finished(&mut self) -> StreamExecutorResult<bool> {
-        // never finish when consuming snapshot
-        Ok(false)
+    fn can_consume_upstream(&self) -> bool {
+        // no backpressure to upstream when consuming snapshot
+        true
     }
 
     fn on_upstream_barrier(&mut self, upstream_barrier: DispatcherBarrier) {
         self.pending_barriers.push(upstream_barrier)
     }
 }
 
-struct StateOfConsumingLogStore<'a> {
-    barrier_rx: &'a mut mpsc::UnboundedReceiver<Barrier>,
+struct StateOfConsumingLogStore {
     /// Barriers received from upstream but not yet received the barrier from local barrier worker
     /// newer barrier at the front
     upstream_pending_barriers: VecDeque<DispatcherBarrier>,
-    /// Barriers received from both upstream and local barrier worker
-    /// newer barrier at the front
-    barriers: VecDeque<Barrier>,
-    is_finished: bool,
-    current_subscriber_id: u32,
-    upstream_table_id: TableId,
+    max_pending_barrier_num: usize,
+    /// The amount of `max_pending_barrier_num` to decrease whenever emitting a checkpoint barrier to downstream
+    max_pending_barrier_num_gap: usize,
 }
 
-impl<'a> StateOfConsumingLogStore<'a> {
+impl StateOfConsumingLogStore {
     fn barrier_count(&self) -> usize {
-        self.upstream_pending_barriers.len() + self.barriers.len()
-    }
-
-    async fn handle_one_pending_barrier(&mut self) -> StreamExecutorResult<Barrier> {
-        assert!(!self.is_finished);
-        let barrier = receive_next_barrier(self.barrier_rx).await?;
-        assert_eq!(
-            self.upstream_pending_barriers
-                .pop_back()
-                .expect("non-empty")
-                .epoch,
-            barrier.epoch
-        );
-        if is_finish_barrier(&barrier, self.current_subscriber_id, self.upstream_table_id) {
-            self.is_finished = true;
-        }
-        Ok(barrier)
+        self.upstream_pending_barriers.len()
     }
 }
 
-impl<'a> UpstreamBufferState for StateOfConsumingLogStore<'a> {
-    async fn is_finished(&mut self) -> StreamExecutorResult<bool> {
-        while !self.upstream_pending_barriers.is_empty() {
-            let barrier = self.handle_one_pending_barrier().await?;
-            self.barriers.push_front(barrier);
-        }
-        if self.is_finished {
-            assert!(self.upstream_pending_barriers.is_empty());
+impl UpstreamBufferState for StateOfConsumingLogStore {
+    fn can_consume_upstream(&self) -> bool {
+        if let Some(prev_barrier) = self.upstream_pending_barriers.front()
+            && prev_barrier.kind.is_barrier()
+        {
+            // allow consuming upstream when the barrier is non-checkpoint barrier to avoid deadlock
```

Comment:
Why deadlock can happen if prev_barrier is checkpoint barrier?

Reply:
Let's say we have epoch 1, 2, 3, where 1, 2 is non-checkpoint epoch, and 3 is checkpoint barrier. After receiving barrier of epoch 1, the snapshot backfill executor will do `iter_log` on epoch 1 to replay the change log, before which it should wait for the committed epoch of upstream to bump to above epoch 1. However, since epoch 1 is a non-checkpoint epoch, we actually need to wait until the upstream to commit epoch 3. But we currently only receive barrier 1 from upstream, and then stop polling, which can have back-pressure on the upstream, and this back pressure may block the upstream from processing epoch 3, and cause the deadlock.

In short, the deadlock will be like, `downstream iter_log on epoch 1` -> `downstream wait on upstream finish epoch 3` -> `upstream blocked by back pressure caused by downstream not polling further barrier` -> `downstream not poll barrier due to being blocking by iter_log on epoch1`

Reply:
Makes sense, thank you